### ISSUE TITLE
Feature request: `from_fn` adapter

### ISSUE THREAD

**Issue by Kmeakin** [jDomantas/internal-iterator#16]
It would be helpful to have an `InternalIterator` version of [`std::iter::from_fn`](https://doc.rust-lang.org/std/iter/fn.from_fn.html)

**Comment by jDomantas** [comment]
Currently this can be achieved by wrapping an std iterator with the adapter: `std::iter::from_fn(...).into_internal()`. And it's as good as a direct implementation would be (which would only give an ergonomics improvement).

I will add this once I find some time.

**Comment by Kmeakin** [comment]
I meant a function which returns `ControlFlow<R>`, so you can write an internal iterator without needing to create a new struct. Something like:

```rust
internal_iterator::from_fn(|f| {
    f(item)?;
    ControlFlow::Continue(())
})
```

I'm not sure if that is possible with Rust's type system

**Comment by jDomantas** [comment]
That would require provided closure to be generic (as its argument `f` is supposed to be `impl FnMut(Item)`), which is currently not possible. There might be a way to do this by making from_fn a macro, or by making the closure take a `&mut dyn FnMut(Item)` instead, I'll see what I can do.

**Comment by Kmeakin** [comment]
You could do a macro that generates a struct and an impl of `InternalIterator` given a body of `try_for_each`

**Comment by jDomantas** [comment]
Provided body would not be able to capture any of its environment, so I doubt the usefulness of such macro (unless you can show me specific examples where you'd like to use it).

I managed to implement a from_fn function that takes a plain closure: [playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=1cb1d8621d6257b88afed16a78a98281). It requires dynamic dispatch and the type signature becomes a bit convoluted which I am not a fan of, but does this look like something that would be useful to you?

**Comment by Kmeakin** [comment]
Hmm, that could work. Depends on how well LLVM is able to devirtualise the indirect call. Maybe include it in the library, but with a warning that it may be slower than a manual implementation?

**Comment by jDomantas** [comment]
I have published 0.2.3, here it is: https://docs.rs/internal-iterator/0.2.3/internal_iterator/fn.from_fn.html
